pkg/vectorize/date_sub/date_sub.go

```diff
@@ -32,76 +32,171 @@ func init() {
+				continue
+			}
+			date, success := d.ToTime().AddInterval(-ys[0], types.IntervalType(zs[0]), types.DateType)
+			if success {
+				rs[i] = date.ToDate()
+			} else {
+				return rs, types.ErrInvalidDateAddInterval
+			}
 		}
 	}
 	return rs, nil
 }
 
-func datetimeSub(xs []types.Datetime, ys []int64, zs []int64, ns *nulls.Nulls, rs []types.Datetime) ([]types.Datetime, error) {
+func datetimeSub(xs []types.Datetime, ys []int64, zs []int64, xns *nulls.Nulls, yns *nulls.Nulls, rns *nulls.Nulls, rs []types.Datetime) ([]types.Datetime, error) {
 	if len(ys) == 0 || len(zs) == 0 {
 		for i := range xs {
-			nulls.Add(ns, uint64(i))
+			nulls.Add(rns, uint64(i))
 		}
 		return rs, nil
 	}
-	for i, d := range xs {
-		date, success := d.AddInterval(-ys[0], types.IntervalType(zs[0]), types.DateTimeType)
-		if success {
-			rs[i] = date
-		} else {
-			rs[i] = 0
-			nulls.Add(ns, uint64(i))
+	if len(xs) == len(ys) {
+		for i, d := range xs {
+			if nulls.Contains(xns, uint64(i)) || nulls.Contains(yns, uint64(i)) {
+				nulls.Add(rns, uint64(i))
+				continue
+			}
+			date, success := d.AddInterval(-ys[i], types.IntervalType(zs[0]), types.DateType)
+			if success {
+				rs[i] = date
+			} else {
+				return rs, types.ErrInvalidDateAddInterval
+			}
+		}
+	} else if len(xs) == 1 {
+		for i, d := range ys {
+			if nulls.Contains(xns, uint64(0)) || nulls.Contains(yns, uint64(i)) {
+				nulls.Add(rns, uint64(i))
+				continue
+			}
+			date, success := xs[0].AddInterval(-d, types.IntervalType(zs[0]), types.DateType)
+			if success {
+				rs[i] = date
+			} else {
+				return rs, types.ErrInvalidDateAddInterval
+			}
+		}
+	} else if len(ys) == 1 {
+		for i, d := range xs {
+			if nulls.Contains(xns, uint64(i)) || nulls.Contains(yns, uint64(0)) {
+				nulls.Add(rns, uint64(i))
+				continue
+			}
+			date, success := d.AddInterval(-ys[0], types.IntervalType(zs[0]), types.DateType)
+			if success {
+				rs[i] = date
+			} else {
+				return rs, types.ErrInvalidDateAddInterval
+			}
 		}
 	}
 	return rs, nil
 }
 
-func dateStringSub(xs *types.Bytes, ys []int64, zs []int64, ns *nulls.Nulls, rs []types.Datetime) ([]types.Datetime, error) {
+func dateStringSub(xs *types.Bytes, ys []int64, zs []int64, xns *nulls.Nulls, yns *nulls.Nulls, rns *nulls.Nulls, rs []types.Datetime) ([]types.Datetime, error) {
 	if len(ys) == 0 || len(zs) == 0 {
 		for i := range xs.Lengths {
-			nulls.Add(ns, uint64(i))
+			nulls.Add(rns, uint64(i))
 			rs[i] = 0
 		}
 		return rs, nil
 	}
+	ds := make([]types.Datetime, len(xs.Lengths))
 	for i := range xs.Lengths {
 		str := string(xs.Get(int64(i)))
 		d, e := types.ParseDatetime(str, 6)
 		if e != nil {
-			// set null
-			nulls.Add(ns, uint64(i))
-			rs[i] = 0
-			continue
+			return rs, e
 		}
-		date, success := d.AddInterval(-ys[0], types.IntervalType(zs[0]), types.DateTimeType)
-		if success {
-			rs[i] = date
-		} else {
-			nulls.Add(ns, uint64(i))
-			rs[i] = 0
+		ds[i] = d
+	}
+	if len(ds) == len(ys) {
```

Comment:
And is this the right test?   Shouldn't that be testing if it is a constant vector then ....